micro_dl/utils/mp_utils.py

```diff
@@ -258,3 +258,52 @@ def resize_and_save(**kwargs):
     )
     # Write image
     cv2.imwrite(kwargs['write_path'], im_resized)
+
+
+def mp_rescale_vol(fn_args, workers):
+    """Rescale and save image stacks with multiprocessing
+
+    :param list of tuple fn_args: list with tuples of function arguments
+    :param int workers: max number of workers
+    """
+
+    with ProcessPoolExecutor(workers) as ex:
+        # can't use map directly as it works only with single arg functions
+        ex.map(rescale_vol_and_save, *zip(*fn_args))
+
+
+def rescale_vol_and_save(time_idx,
+                         pos_idx,
+                         channel_idx,
+                         sl_start_idx,
+                         sl_end_idx,
+                         frames_metadata,
+                         output_fname,
+                         scale_factor,
+                         input_dir):
+    """Rescale volumes and save
+
+    :param int time_idx: time point of input image
+    :param int pos_idx: sample idx of input image
+    :param int channel_idx: channel idx of input image
+    :param int sl_start_idx: start slice idx for the vol to be saved
+    :param int sl_end_idx: end slice idx for the vol to be saved
+    :param pd.Dataframe frames_metadata: metadata for the input slices
+    :param str output_fname: output_fname
+    :param float/list scale_factor: scale factor for resizing
+    :param str input_dir: input dir for 2D images
+    """
+
+    input_stack = []
+    for sl_idx in range(sl_start_idx, sl_end_idx):
+        meta_idx = aux_utils.get_meta_idx(frames_metadata,
+                                          time_idx,
+                                          channel_idx,
+                                          sl_idx,
+                                          pos_idx)
+        cur_fname = frames_metadata.loc[meta_idx, 'file_name']
+        cur_img = image_utils.read_image(os.path.join(input_dir, cur_fname))
+        input_stack.append(cur_img)
+    input_stack = np.stack(input_stack, axis=0)
+    resc_vol = image_utils.rescale_nd_image(input_stack, scale_factor)
```

Comment:
I was just wondering - how will these resizes volumes work with the remainder of the preprocessing - flatfield, masks and tiles (which I think assumes 2D frames)?

Reply:
I've been debating the same! The options I'm thinking are:
1) after resizing, we could save it as 2D slices and the slice_idx will correspond to the center slice but tiling has to be modified to allow even number of slices
2) Save it as 3D volume where slice_idx will correspond to the starting slice in sub volume and use tile 3D option
I've a slight preference for 2 but not too attached to it!